### PR TITLE
compressor: simplify logic

### PR DESCRIPTION
**Pull request by rgs1**
This confused me the first time I read it, so make it simpler
for the next reader -- there's no need to handle `not_compressed_.inc()`
from the two different places. Delay it until encodeHeaders() and
do it once.

Signed-off-by: Raul Gutierrez Segales <rgs@pinterest.com>

### CODE REVIEW

**Comment by rgs1** [comment]
@mattklein123 @dio take 2 :-)

**Comment by rgs1** [comment]
Hmm, failed test passes locally for me:

```
//test/extensions/filters/http/cache:cache_filter_integration_test       PASSED in 4.3s
```

So guessing transient:

```
test/extensions/filters/http/cache/cache_filter_integration_test.cc:68
Value of: request->headers()
Expected: is a superset of headers:
':status', '200'
'date', 'Thu, 01 Jan 1970 01:00:01 GMT'
'cache-control', 'public,max-age=3600'
'content-length', '42'

  Actual: 16-byte object <E0-AB 68-01 00-00 00-00 B0-B3 68-01 00-00 00-00> (of type Envoy::Http::ResponseHeaderMap), where the following matchers don't match any elements:
matcher #1: is equal to ("date", "Thu, 01 Jan 1970 01:00:01 GMT")
Stack trace:
  0x84d18c: Envoy::Extensions::HttpFilters::Cache::CacheIntegrationTest_MissInsertHit_Test::TestBody()
  0x15f1028: testing::internal::HandleExceptionsInMethodIfSupported<>()
  0x15f0f55: testing::Test::Run()
  0x15f1e20: testing::TestInfo::Run()
... Google Test internal frames ...
```

cc: @toddmgreer

**Comment by dio** [comment]
@rojkov could you help to take a look at this? Thanks!

File: test/extensions/filters/http/common/compressor/compressor_filter_test.cc [hunk 1]
```diff
@@ -174,7 +174,7 @@ TEST_F(CompressorFilterTest, DecodeHeadersWithRuntimeDisabled) {
   }
 }
 )EOF");
-  EXPECT_CALL(runtime_.snapshot_, getBoolean("foo_key", true)).WillOnce(Return(false));
+  EXPECT_CALL(runtime_.snapshot_, getBoolean("foo_key", true)).WillRepeatedly(Return(false));
```
Comment (rojkov) [inline test/extensions/filters/http/common/compressor/compressor_filter_test.cc]: Nit: perhaps makes sense to have this expectation as
```
EXPECT_CALL(runtime_.snapshot_, getBoolean("foo_key", true)).Times(2).WillRepeatedly(Return(false));
```

to keep the number of `config_->enabled()` calls under control.

**Pull request merged**